internal/providers/packet/packet.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packet
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/coreos/coreos-metadata/internal/providers"
+	"github.com/coreos/coreos-metadata/internal/retry"
+)
+
+const metadataURL = "https://metadata.packet.net/metadata"
+
+var errNotFound = errors.New("Coudn't find an IP with requested attributes")
+
+type packetMetadata struct {
+	Hostname string        `json:"hostname"`
+	Network  packetNetwork `json:"network"`
+	SSHKeys  []string      `json:"ssh_keys"`
+}
+
+type packetNetwork struct {
+	Addresses []packetAddress `json:"addresses"`
+}
+
+type packetAddress struct {
+	Family  int    `json:"address_family"`
+	Address string `json:"address"`
+	Public  bool   `json:"public"`
+}
+
+func (p *packetMetadata) findIP(family int, public bool) (net.IP, error) {
+	for _, addr := range p.Network.Addresses {
+		if addr.Public == public && addr.Family == family {
+			ip := net.ParseIP(addr.Address)
+
+			if ip == nil {
+				return nil, fmt.Errorf("couldn't parse %q as IP address", addr.Address)
+			}
+
+			return ip, nil
+		}
+	}
+
+	return nil, errNotFound
+}
+
+func (p *packetMetadata) publicIPv4() (net.IP, error) {
+	ip, err := p.findIP(4, true)
+
```

Comment:
Throughout this listing there are newlines separating assignments from their error tests, which is inconsistent with the rest of the code in coreos-metadata.  Please remove these excess newlines.